Sprint 16

desikvasya opened the pull request:
Экран со списком категорий переписан на архитектуру MVVM.
Сделаны экраны онбординга с использованием UIPageViewController.
Добавлен экран с созданием категории, категории отображаются в списке сразу же после создания.

desikvasya committed:
добавил онбординг

desikvasya committed:
Добавлен показ онбординга при первом запуске
Обновление коллекции при запуске

desikvasya committed:
Исправил ошибку
Теперь при создании трекера, он сразу же отображается на главном экране

desikvasya committed:
Переписал экран выбора категорий на MVVM

desikvasya committed:
Убрал сильные ссылки

desikvasya committed:
Добавил создание собственной категории

desikvasya committed:
Созданные категории сразу же отображаются в списке категорий

protos17 commented on Habit/AppData/Data.swift:
```diff
@@ -36,12 +36,12 @@ let colorCollectionData = [
+//]
+//
+//// MARK: - Переменная, хранящая выбранную пользователем категорию события
+//var categoryName = ""
```
Лишний закомментированный код рекомендую удалять

protos17 commented on Habit/CoreData/TrackerCategoryStore.swift:
```diff
@@ -10,8 +10,43 @@ import CoreData
     
-    // MARK: - Method to add a category in the database
-    func addCategory(category: String, tracker: TrackerCoreData, context: NSManagedObjectContext) {
+    private var categories = UserDefaults.standard.array(forKey: "category_list") as? [String]
```
Рекомендую UserDefaults.standard сделать приватной константой здесь, а "category_list" указать в энаме с ключами для UserDefaults

protos17 commented on Habit/CoreData/TrackerCategoryStore.swift:
```diff
@@ -10,8 +10,43 @@ import CoreData
+        return true
+    }
+    
+    func getCategories() -> [String] {
```
Методы, которые не принимают аргументов и возвращaют значение, лучше делать вычисляемым свойством:

    var savedCategories: [String] {
        categories ?? []
    }

protos17 commented on Habit/CoreData/TrackerCategoryStore.swift:
```diff
@@ -10,8 +10,43 @@ import CoreData
+        return index
+    }
+    
+    func getChoosedCategory() -> String {
```
Рекомендую сделать вычисляемым свойством

protos17 commented on Habit/CoreData/TrackerCategoryStore.swift:
```diff
@@ -10,8 +10,43 @@ import CoreData
+    func addCategory(newCategory: String) -> Bool {
+        categories?.append(newCategory)
+        UserDefaults.standard.set(categories, forKey: "category_list")
+        if UserDefaults.standard.synchronize() {
+            return true
+        } else {
+            return false
+        }
```
Здесь можно написать просто:

return UserDefaults.standard.synchronize()

protos17 commented on Habit/TabBarControllers/MainTabBarViewController.swift:
```diff
@@ -9,10 +9,18 @@ import UIKit
 
 final class MainTabBarViewController: UITabBarController {
     
+        var onboarding: UIViewController?
```
Здесь стоит добавить `private`
Необходимо указывать private  для того, чтобы спрятать детали реализации конкретной части функциональности, когда они используются только внутри области объявления.

protos17 commented on Habit/TabBarControllers/MainTabBarViewController.swift:
```diff
@@ -33,4 +41,14 @@ final class MainTabBarViewController: UITabBarController {
     }
+    
+    private func setupProperties() {
+         UserDefaults.standard.set(true, forKey: "isLogged")
+         let categoryList = UserDefaults.standard.array(forKey: "category_list") as? [String]
```
Работу с UserDefaults рекомендую реализовать в отдельном менеджере, как это делали в предыдущем проекте с картинками

protos17 commented on Habit/TrackerCreation/CategorySelectionViewController.swift:
```diff
@@ -9,6 +9,17 @@ import UIKit
 
 final class CategorySelectionViewController: UIViewController {
     
+    let viewModel: CategoryViewModel
```
Рекомендую указать как private

protos17 commented on Habit/TrackerCreation/CategorySelectionViewController.swift:
```diff
@@ -65,18 +76,20 @@ final class CategorySelectionViewController: UIViewController {
     
     // MARK: - Настройка внешнего вида
     private func setupView() {
+        NotificationCenter.default.addObserver(self, selector: #selector(updateCategories), name: Notification.Name("categories_added"), object: nil)
```
Рекомендую добавить еще отписку при закрытии экрана

protos17 commented on Habit/TrackerCreation/CategorySelectionViewController.swift:
```diff
@@ -95,27 +108,71 @@ final class CategorySelectionViewController: UIViewController {
+            categoriesTable.isHidden = false
+        } else {
+            stackView.isHidden = false
+            categoriesTable.isHidden = true
```
Рекомендую заменить на:

stackView.isHidden = !viewModel.getCategories().isEmpty
categoriesTable.isHidden = viewModel.getCategories().isEmpty

protos17 commented on Habit/TrackerCreation/NewCategoryViewController.swift:
```diff
@@ -0,0 +1,147 @@
+final class NewCategoryViewController: UIViewController {
+    
+    // MARK: - Свойства
+    let categoryViewModel: CategoryViewModel
```
Рекомендую указать как private

protos17 commented on Habit/TrackerCreation/NewCategoryViewController.swift:
```diff
@@ -0,0 +1,147 @@
+        return button
+    }()
+    
+    let enterNameTextField: UITextField = {
```
Рекомендую указать как private

protos17 commented on Habit/TrackerCreation/NewCategoryViewController.swift:
```diff
@@ -0,0 +1,147 @@
+            case true: self.dismiss(animated: true)
+                let notification = Notification(name: Notification.Name("categories_added"))
+                NotificationCenter.default.post(notification)
+            case false: print("Ошибка добавления категории")
```
При ошибке рекомендую отображать алерт с ошибкой, как мы это делали в предыдущих проектах

protos17 commented on Habit/TrackerCreation/NewHabitViewController.swift:
```diff
@@ -8,8 +8,21 @@
+    
     let dataProvider = DataProvider()
     
+    let categoryViewModel: CategoryViewModel
```
Рекомендую указать как private

protos17 commented on Habit/TrackerCreation/NewHabitViewController.swift:
```diff
@@ -8,8 +8,21 @@
+    }
+
     // MARK: - Свойства
     let titleLabel: UILabel = {
```
UI элементы также рекомендую указывать как private

protos17 commented on Habit/TrackerCreation/NewIrregularEventViewController.swift:
```diff
@@ -11,6 +11,17 @@ final class NewIrregularEventViewController: UIViewController {
     
     let dataProvider = DataProvider()
     
+    let categoryViewModel: CategoryViewModel
```
Рекомендую указать как private

protos17 commented on Habit/TrackerCreation/TrackerTypeViewController.swift:
```diff
@@ -10,6 +10,8 @@ import UIKit
+    let categoryViewModel = CategoryViewModel()
+    
     // MARK: - Свойства
     let titleLabel: UILabel = {
```
Рекомендую указать как private

protos17 commented on Habit/ViewModels/CategoryViewModel.swift:
```diff
@@ -0,0 +1,48 @@
+        isCategoryChoosed?(result)
+    }
+    
+    func getCategories() -> [String] {
```
Рекомендую сделать вычисляемым свойством:

var categorise: [String] {
model.getCategories()
}

protos17 commented on Habit/ViewModels/CategoryViewModel.swift:
```diff
@@ -0,0 +1,48 @@
+    
+    func getChoosedCategory() -> String {
+        return model.getChoosedCategory()
+    }
```
Также рекомендую указать вычисляемым свойством

protos17 commented on Habit/ViewModels/CategoryViewModel.swift:
```diff
@@ -0,0 +1,48 @@
+
+typealias Binding<T> = (T) -> Void
+
+final class CategoryViewModel {
```
Рекомендую сделать протокол и все публичные методы указать в нем

protos17 commented on Habit/TrackerCreation/CategorySelectionViewController.swift:
```diff
@@ -95,27 +108,71 @@ final class CategorySelectionViewController: UIViewController {
+            guard let self = self else { return }
+            if isOk {
+                self.dismiss(animated: true)
+                let notification = Notification(name: Notification.Name("category_changed"))
```
Рекомендую имена для NotificationCenter указать константами, чтобы можно было обращаться к ним как к энаму, что избавит от возможных ошибок при случайной опечатки в имени нотификации:

extension Notification.Name {
    static let categoriesDidChanged = Notification.Name("category_changed")
}

и заменить на:

                let notification = Notification(name: .categoriesDidChanged)

то же самое сделать для остальных Notification.Name

desikvasya committed:
Исправления после код-ревью:
- заменил функции на вычисляемые значения
- косметические изменения
- добавил private, где было необходимо
- улучшил работу с нотификациями
